File: fsharp/BronKerbosch.Test/DegeneracyTest.fs
module BronKerbosch.Test.Degeneracy

open NUnit.Framework
open BronKerbosch
open BronKerbosch.Test.UndirectedGraph

[<Test>]
let order0 () =
    let g = graph_order0
    Assert.That(Degeneracy.Of(g).iter (), Is.Empty)

[<Test>]
let order1 () =
    let g = graph_order1
    Assert.That(Degeneracy.Of(g).iter (), Is.Empty)

[<Test>]
let order2_isolated () =
    let g = graph_order2_isolated
    Assert.That(Degeneracy.Of(g).iter (), Is.Empty)

[<Test>]
let order2_connected () =
    let g = graph_order2_connected
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 1)

[<Test>]
let order3_size1_left () =
    let g = graph_order3_size1_left
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 1)

[<Test>]
let order3_size1_long () =
    let g = graph_order3_size1_long
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 1)

[<Test>]
let order3_size1_right () =
    let g = graph_order3_size1_right
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 1)

[<Test>]
let order3_size2 () =
    let g = graph_order3_size2
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 2)

[<Test>]
let order3_size3 () =
    let g = graph_order3_size3
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 2)

[<Test>]
let order4_size2 () =
    let g = graph_order4_size2
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 2)

[<Test>]
let order4_size3_bus () =
    let g = graph_order4_size3_bus
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 3)
    Assert.That(Degeneracy.Of(g).iter () |> Seq.head, Is.AnyOf [| Vertex 0; Vertex 3 |])

[<Test>]
let order4_size3_star () =
    let g = graph_order4_size3_star
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 3)
    Assert.That(Degeneracy.Of(g).iter () |> Seq.head, Is.Not.EqualTo(Vertex 0))

[<Test>]
let order4_size4_p () =
    let g = graph_order4_size4_p
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 3)
    Assert.That(Degeneracy.Of(g).iter () |> Seq.head, Is.EqualTo(Vertex 0))

[<Test>]
let order4_size4_square () =
    let g = graph_order4_size4_square
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 3)

[<Test>]
let order4_size5 () =
    let g = graph_order4_size5
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 3)
    Assert.That(Degeneracy.Of(g).iter () |> Seq.head, Is.AnyOf [| Vertex 1; Vertex 3 |])

[<Test>]
let order5_size6_penultimate () =
    let g = graph_order5_size6_penultimate
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 4)

[<Test>]
let sample () =
    let g = graph_sample
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 6)
    Assert.That(Degeneracy.Of(g).iter () |> Seq.head, Is.AnyOf [| Vertex 6; Vertex 7 |])

[<Test>]
let bigger () =
    let g = graph_bigger
    Assert.That(Degeneracy.Of(g).iter () |> Seq.length, Is.EqualTo 9)
    Assert.That(Degeneracy.Of(g).iter () |> Seq.head, Is.AnyOf [| Vertex 5; Vertex 8 |])

File: csharp/BronKerbosch.Test/DegeneracyTest.cs
using NUnit.Framework;
using System.Collections.Generic;
using System.Linq;

namespace BronKerbosch.Test
{
    public class DegeneracyHashTest : DegeneracyTestTemplate<HashSet<Vertex>, HashSetMgr> { }
    public class DegeneracySortedTest : DegeneracyTestTemplate<SortedSet<Vertex>, SortedSetMgr> { }

    public class DegeneracyTestTemplate<TVertexSet, TVertexSetMgr> : LabGraphs<TVertexSet, TVertexSetMgr>
        where TVertexSet : ISet<Vertex>
        where TVertexSetMgr : IVertexSetMgr<TVertexSet>
    {
        [Test]
        public void TestOrder0()
        {
            var g = Order0.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g), Is.Empty);
        }

        [Test]
        public void TestOrder1()
        {
            var g = Order1.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g), Is.Empty);
        }

        [Test]
        public void TestOrder2_isolated()
        {
            var g = Order2_isolated.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g), Is.Empty);
        }

        [Test]
        public void TestOrder2_connected()
        {
            var g = Order2_connected.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(1));
        }

        [Test]
        public void TestOrder3_size1_left()
        {
            var g = Order3_size1_left.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(1));
        }

        [Test]
        public void TestOrder3_size1_long()
        {
            var g = Order3_size1_long.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(1));
        }

        [Test]
        public void TestOrder3_size1_right()
        {
            var g = Order3_size1_right.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(1));
        }

        [Test]
        public void TestOrder3_size2()
        {
            var g = Order3_size2.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(2));
        }

        [Test]
        public void TestOrder3_size3()
        {
            var g = Order3_size3.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(2));
        }

        [Test]
        public void TestOrder4_size2()
        {
            var g = Order4_size2.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(2));
        }

        [Test]
        public void TestOrder4_size3_bus()
        {
            var g = Order4_size3_bus.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(3));
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Select((pair) => pair.Item1).First(), Is.AnyOf([Vertex.Nth(0), Vertex.Nth(3)]));
        }

        [Test]
        public void TestOrder4_size3_star()
        {
            var g = Order4_size3_star.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(3));
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Select((pair) => pair.Item1).First, Is.Not.EqualTo(Vertex.Nth(0)));
        }

        [Test]
        public void TestOrder4_size4_p()
        {
            var g = Order4_size4_p.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(3));
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Select((pair) => pair.Item1).First, Is.EqualTo(Vertex.Nth(0)));
        }

        [Test]
        public void TestOrder4_size4_square()
        {
            var g = Order4_size4_square.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(3));
        }

        [Test]
        public void TestOrder4_size5()
        {
            var g = Order4_size5.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(3));
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Select((pair) => pair.Item1).First, Is.AnyOf([Vertex.Nth(1), Vertex.Nth(3)]));
        }

        [Test]
        public void TestOrder5_size6_penultimate()
        {
            var g = Order5_size6_penultimate.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(4));
        }

        [Test]
        public void TestSample()
        {
            var g = Sample.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(6));
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Select((pair) => pair.Item1).First, Is.AnyOf([Vertex.Nth(6), Vertex.Nth(7)]));
        }

        [Test]
        public void TestBigger()
        {
            var g = Bigger.Graph;
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Count, Is.EqualTo(9));
            Assert.That(Degeneracy<TVertexSet, TVertexSetMgr>.Iter(g).Select((pair) => pair.Item1).First, Is.AnyOf([Vertex.Nth(5), Vertex.Nth(8)]));
        }
    }
}
